paddle/pybind/pybind.cc

```diff
@@ -25,9 +26,143 @@ namespace pd = paddle::framework;
 
 USE_OP(add_two);
 
+struct PlaceDebugString : public boost::static_visitor<std::string> {
+  std::string operator()(const paddle::platform::GPUPlace& place) const {
+    return "GPU(" + std::to_string(place.device) + ")";
+  }
+
+  std::string operator()(const paddle::platform::CPUPlace& place) const {
+    return "CPU";
+  }
+};
+
+template <typename T>
```

Comment:
## Why use `bool less` in template
The `bool less` is useful, because C++ 11 lack of `static if` statement, all codes in that block will be compiled, and part of them cannot be specialized. 

```cpp
std::tuple_element<TypeIndex, Types>::type;
```

this line will cause a compile error because TypeIndex might larger than tuple size.

The template argument `less` is used for specializing two conditions. When `less=False`, we do not need to get the `current type` from the tuple.

Another way is to use `std::enable_if` to do the similar thing. But that will be harder to read.

## Why use `functor`, i.e, `struct with operator()`, instead of `function`.

Because `partial specialization is not allowed` for cpp function. e.g. That code segment will never be compiled by a CPP compiler.

```cpp
template <bool flag, typename T>
int foo();

template <typename T>
T foo<true, T>() {
  return 0;
}

template <typename T>
T foo<false, T>() {
  return 0;
}
```